### DISCUSSION TITLE
Reading historical data with opcua-asyncio client in synchronous code

### DISCUSSION THREAD

**Discussion by KallistoD** [FreeOpcUa/opcua-asyncio#1345] [Q&A] [answered]
As written on [opcua-asyncio library home page](https://github.com/FreeOpcUa/opcua-asyncio), there is a [synchronous wrapper](https://github.com/FreeOpcUa/opcua-asyncio/blob/master/asyncua/sync.py) over the async API, which can be used in synchronous code. But there are no methods in synchronous wrapper for historical data read. It raises exception like `'SyncNode' object has no attribute 'read_raw_history'` if attempting to use such methods.
I also tried to use async Client from synchronous code like this:
`from asyncua import Client`
`client = asyncio.run(Client(url=server_url))`
But seems, this approach cannot be applied for async constructors and it does not work actually.
Is there a way to use opcua-asyncio library to access historical data from synchronous code, please? Maybe, there is any workaround? In my case, I have CPU-bound historical data processing, which is separated from GUI process.

**Comment by schroeder-** [comment, marked as answer]
Adding new function is easy see pr #515 for example. A pr for this is always welcome.

**Comment by KallistoD** [comment]
Understood. Thank you.

**Reply by KallistoD** [reply]
[Pull request](https://github.com/FreeOpcUa/opcua-asyncio/pull/1349) for the reference.
